patches/server/0988-fix-MC-65198-ITEM_CRAFTED-statistic-isn-t-always-inc.patch

```diff
@@ -0,0 +1,56 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: elmital <54907162+elmital@users.noreply.github.com>
+Date: Fri, 14 Jul 2023 18:17:38 +0200
+Subject: [PATCH] fix: MC-65198 ITEM_CRAFTED statistic isn't always incremented
+ correctly
+
+
+diff --git a/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java b/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
+index ff770b9ce68a62418de0c7ed389650626fa1dcb2..e939bbc42216d688be655c47fff362906ed5f3f4 100644
+--- a/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
++++ b/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
+@@ -73,6 +73,13 @@ public abstract class ItemCombinerMenu extends AbstractContainerMenu {
+             public void onTake(Player player, ItemStack stack) {
+                 ItemCombinerMenu.this.onTake(player, stack);
+             }
++
++            // Paper start - fix MC-65198 shift clicking in result slot of a smiting table
++            @Override
++            protected void onQuickCraft(ItemStack stack, int amount) {
++                stack.onCraftedBy(player.level(), player, amount);
++            }
++            // Paper end
+         });
+     }
+ 
+diff --git a/src/main/java/net/minecraft/world/inventory/ResultSlot.java b/src/main/java/net/minecraft/world/inventory/ResultSlot.java
+index ddf87ef1979c4ae4f14cd6aa220f9146d5e20909..4a8529ebb1bfce015bd77a35076fcc3bee33fa90 100644
+--- a/src/main/java/net/minecraft/world/inventory/ResultSlot.java
++++ b/src/main/java/net/minecraft/world/inventory/ResultSlot.java
+@@ -45,7 +45,7 @@ public class ResultSlot extends Slot {
+     @Override
+     protected void checkTakeAchievements(ItemStack stack) {
+         if (this.removeCount > 0) {
+-            stack.onCraftedBy(this.player.level(), this.player, this.removeCount);
++            stack.onCraftedBy(this.player.level(), this.player, stack.getCount()); // Paper - fix MC-65198 for drop action in result slot
```

Comment:
This removes the use of the `removeCount` field entirely. I cannot see a use case where it is actually needed, but I think caution should be taken if we just do this without considering what the field is actually used for, or why it exists.

For example, the `removeCount` field is absolutely required for the `FurnaceResultSlot` which is largely based on ResultSlot. However, it seems everywhere ResultSlot is used, the whole stack in the slot is removed always which would mean using stack is fine here. But we need to be sure.

Reply:
I haven't really found an use of the field either excluding like you said in the `FurnaceResultSlot`. But without be certain I prefered to let it like this. The whole stuff is a little bit messy from my point of view for example the crafting statistic increase methods are called twice. In the past Mojang has already fix [issue](https://bugs.mojang.com/browse/MC-29515) with statistics incrementation so maybe the field should be a remainder of that? The incrementation is called a first time with a positive count value then after called with a 0 count value so the `removeCount` field check in `ResultSlot#CheckTakeAchievements(ItemStack)` was may be at first here to prevent the increase of the statistic twice idk. But since the `ItemStack#getCount()` is already set to 0 at the second call even without the `removeCount` check it should not increase the Statistic now. I didn't want to dive more into the stuff cause It could need a bigger refactor of the entire crafting stuff for the same result without known what Mojang planned for this part with all new job blocks.